### ISSUE TITLE
Fix S2583: Rule should not report false positives when combined with async methods

### ISSUE THREAD

**Issue by shadow-cs** [SonarSource/sonar-csharp#1400]
### Description

Using `await` should stop detection of the same condition checks if a variable (field) from outer scope is used in the condition.

### Repro steps

```csharp
class FooBar
{
  int foo = 0;

  public async Task Bar()
  {
    var i = ++foo;
    await Task.Delay(100);
    if (i != foo)
      return;
    await Task.Delay(100);
    if (i != foo) // <-- S2583 reported here
      return;
    Baz();
  }

  void Baz() {}
}
```

Even though method from another class is called in the example (and without async operation it cannot change the `foo` field) it might get changed before the method resumes asynchronously (`foo` might get changed possibly by calling `Bar` again before the delay has finished).

### Expected behavior

No warnings should be generated.

### Actual behavior

Following warning is generated:

> Warning S2583 Change this condition so that it does not always evaluate to 'false'; some subsequent code is never executed.

### Known workarounds

Add a suppression.

### Related information

* SonarC# 4.0.0.3479
* Visual Studio 15.7.1
